src/compiler/performanceCore.ts

```diff
@@ -0,0 +1,80 @@
+/*@internal*/
+namespace ts {
+    export interface PerformanceHooks {
+        performance: Performance;
+        PerformanceObserver: PerformanceObserverConstructor;
+    }
+
+    export interface Performance {
+        clearMarks(name?: string): void;
+        mark(name: string): void;
+        measure(name: string, startMark?: string, endMark?: string): void;
+        now(): number;
+        timeOrigin: number;
+    }
+
+    export interface PerformanceEntry {
+        name: string;
+        entryType: string;
+        startTime: number;
+        duration: number;
+    }
+
+    export interface PerformanceObserverEntryList {
+        getEntries(): PerformanceEntryList;
+        getEntriesByName(name: string, type?: string): PerformanceEntryList;
+        getEntriesByType(type: string): PerformanceEntryList;
+    }
+
+    export interface PerformanceObserver {
+        disconnect(): void;
+        observe(options: { entryTypes: readonly string[] }): void;
+    }
+
+    export type PerformanceObserverConstructor = new (callback: (list: PerformanceObserverEntryList, observer: PerformanceObserver) => void) => PerformanceObserver;
+    export type PerformanceEntryList = PerformanceEntry[];
+
+    declare const performance: Performance | undefined;
+    declare const PerformanceObserver: PerformanceObserverConstructor | undefined;
+
+    function tryGetWebPerformanceHooks(): PerformanceHooks | undefined {
+        if (typeof performance === "object" &&
+            typeof performance.timeOrigin === "number" &&
+            typeof performance.clearMarks === "function" &&
+            typeof performance.mark === "function" &&
+            typeof performance.measure === "function" &&
+            typeof performance.now === "function" &&
+            typeof PerformanceObserver === "function") {
+            return {
+                performance,
+                PerformanceObserver
+            };
+        }
+    }
+
+    function tryGetNodePerformanceHooks(): PerformanceHooks | undefined {
+        if (typeof module === "object" && typeof require === "function") {
+            try {
+                return require("perf_hooks") as typeof import("perf_hooks");

```

Comment:
The conversion is so that it will break if node drops an API we depend on?

Reply:
Essentially, yes. If they change the API in a breaking way (assuming we update the node types in accordance with that change). Also ensures that the `PerformanceHooks` and related interfaces are compatible.